components/director/internal/domain/runtime/service.go

```diff
@@ -377,42 +369,30 @@ func (s *service) Delete(ctx context.Context, id string) error {
 	return nil
 }
 
-// SetLabel missing godoc
+// SetLabel sets Runtime label from a given input
 func (s *service) SetLabel(ctx context.Context, labelInput *model.LabelInput) error {
 	rtmTenant, err := tenant.LoadFromContext(ctx)
 	if err != nil {
 		return errors.Wrapf(err, "while loading tenant from context")
 	}
 
-	err = s.ensureRuntimeExists(ctx, rtmTenant, labelInput.ObjectID)
-	if err != nil {
+	if err = s.ensureRuntimeExists(ctx, rtmTenant, labelInput.ObjectID); err != nil {
 		return err
 	}
 
-	currentRuntimeLabels, err := s.getCurrentLabelsForRuntime(ctx, rtmTenant, labelInput.ObjectID)
-	if err != nil {
+	if modifiable, err := isLabelModifiable(labelInput.Key, s.protectedLabelPattern, s.immutableLabelPattern); err != nil {
 		return err
+	} else if !modifiable {
+		return apperrors.NewInvalidDataError("could not set unmodifiable label with key %s", labelInput.Key)
 	}
 
-	newRuntimeLabels := make(map[string]interface{})
-	for k, v := range currentRuntimeLabels {
-		newRuntimeLabels[k] = v
-	}
-
-	newRuntimeLabels[labelInput.Key] = labelInput.Value
-
-	if err = s.upsertScenariosLabelIfShould(ctx, labelInput.ObjectID, labelInput.Key, newRuntimeLabels); err != nil {
-		return err
-	}
+	id := labelInput.ObjectID
 
-	modifiable, err := isLabelModifiable(labelInput.Key, s.protectedLabelPattern, s.immutableLabelPattern)
-	if err != nil {
-		return err
-	}
-	if !modifiable {
-		return apperrors.NewInvalidDataError("could not set unmodifiable label with key %s", labelInput.Key)
-	}
-	if labelInput.Key != model.ScenariosKey {
+	if labelInput.Key == model.ScenariosKey {
+		if err := s.updateScenariosLabel(ctx, rtmTenant, id, map[string]interface{}{model.ScenariosKey: labelInput.Value}); err != nil {
```

Comment:
The `updateScenariosLabel` seems to work well for the `Update` method but I'm not sure about the `SetLabel` method. What is my concern:
From the old code, we have LabelInput with the new label value that is added to the `currentRuntimeLabels` resulting in `newRuntimeLabels`, and then `s.upsertScenariosLabelIfShould` was invoked with it(along with other parameters). Then `s.upsertScenariosLabelIfShould ` merge the provided `newRuntimeLabels` with those from the ASA and Upsert/Update them.

Now with the new changes, if I understand it correctly, we call `updateScenariosLabel` if the input label key is `scenarios` and provide the label value to the method. `updateScenariosLabel ` then get this input and merge it with the "scenarios from the ASA", called `mergedScenariosSlice` that will be used to invoke `AssignFormation` and for the other scenarios(`currentScenarios`), `UnassignFormation` will be invoked.

To me, it seems to have two different behaviors? Example:
Consider the following setup,
runtime with two scenarios(`scenario1` and `scenario2`) + ASA that will add another scenario - `asa-scenario`. And label value input - `input-label-val` as argument of the `SetLabel` method.
The "old SetLabel behavior" will get the `scenario1` and `scenario2` will add `input-label-val`, then call `upsertScenariosLabelIfShould` which will add the scenarios from the ASA - `asa-scenario` and finally will execute upsert, resulting in "scenarios": ["scenario1", "scenario2", "input-label-val", "asa-scenario"]

The "new SetLabel behavior" will call `updateScenariosLabel` with the label input value, provided to the `SetLabel` method - `input-label-val`, and from the explanation above for the `updateScenariosLabel ` to me it seems to be result in: "scenarios": ["input-label-val", "asa-scenario"]? The other two scenarios will be part of the `currentScenariosSlice` and call `UnassignFormation` for them? 

Am I missing something? I could be totally wrong, that's why I describe it in details. If needed we could have a quick conversation/sync

Reply:
We've discussed it and also tested the example above with the 'old SetLabel behavior' and the result was `scenarios": ["input-label-val", "asa-scenario"]`.